src/families.h

```diff
@@ -18,139 +18,106 @@
-    y_scale.resize(n_classes, 1.0);
-    y_mat.setZero(n_samples, n_classes);
-    y_mat_scale.resize(n_classes, 1.0);
-  };
-
-  virtual double Link(const unsigned i) = 0;
-
-  virtual double Loss(const double prediction, const unsigned i) = 0;
-
-  virtual double Loss(const std::vector<double>& prediction,
-                      const unsigned i) = 0;
-
-  virtual void Gradient(std::vector<double>&       g_i,
-                        const std::vector<double>& prediction,
-                        const unsigned             i) = 0;
-
-  virtual void PreprocessResponse() = 0;
-
-  virtual double NullDeviance(const bool fit_intercept) = 0;
-
-  std::vector<double> StepSize(const double               max_squared_sum,
-                               const std::vector<double>& alpha_scaled,
-                               const bool                 fit_intercept) {
-    // Lipschitz constant approximation
-    std::vector<double> step_sizes;
-    step_sizes.reserve(alpha_scaled.size());
-
-    for (auto alpha_i : alpha_scaled) {
-      double L =
-        L_scaling*(max_squared_sum + static_cast<double>(fit_intercept))
-        + alpha_i;
-      double mu_n = 2.0*n_samples*alpha_i;
-      step_sizes.push_back(1.0 / (2.0*L + std::min(L, mu_n)));
-    }
-    return step_sizes;
-  }
+  Family(const double L_scaling) : L_scaling(L_scaling) {};
+
+  void Preprocess(std::vector<double>& y,
+                  std::vector<double>& y_center,
+                  std::vector<double>& y_scale) noexcept;
+
+  double Loss(const std::vector<double>& prediction,
+              const std::vector<double>& y,
+              const unsigned             i) const noexcept;
+
+  void Gradient(const std::vector<double>& prediction,
+                const std::vector<double>& y,
+                const unsigned             i,
+                std::vector<double>&       gradient) const noexcept;
+
+  double NullDeviance(const std::vector<double>& y,
+                      const bool                 fit_intercept,
+                      const unsigned             n_classes) const noexcept;
+
+  Eigen::MatrixXd
+  LambdaResponse(const std::vector<double>& y,
+                 std::vector<double>&       y_mat_scale) const noexcept;
 
-  std::vector<double> y;
-  std::vector<double> y_center;
-  std::vector<double> y_scale;
-  Eigen::MatrixXd y_mat;
-  std::vector<double> y_mat_scale;
-  unsigned n_samples;
-  unsigned n_classes;
-  unsigned n_targets;
-  double null_deviance = 0.0;
   double L_scaling;
-  double lambda_scaling = 1.0;
 };
 
 class Gaussian : public Family {
 public:
-  Gaussian(const std::vector<double>& y,
-           const unsigned n_samples,
-           const unsigned n_classes)
-           : Family(y, n_samples, n_classes, 1.0) {}
+  Gaussian() : Family(1.0) {}
 
-  void PreprocessResponse() {
+  void Preprocess(std::vector<double>& y,
+                  std::vector<double>& y_center,
+                  std::vector<double>& y_scale) noexcept {
```

Comment:
Using `y_center` and `y_scale` as `std::vector` here leads to a pointer dereference that might be unnecessary in the univariate case. Is it worth templating / breaking `Family` into two cases - `UnivariateFamily` and `MultivariateFamily` to avoid this?

Reply:
Such a split would also let you use POD `double`s in predict, loss, etc.

Reply:
It is a good idea and I have considered it. However, there is no performance gain to be had regarding the y_scale and y_center vectors since they are not used inside the algorithm (where the real bottleneck is).